### ISSUE TITLE
Index watcher triggers on detached buffers

### ISSUE THREAD

**Issue by rgrinberg** [lewis6991/gitsigns.nvim#138]
I sometimes get teh following error:
```
|| ...acker/start/plenary.nvim/lua/plenary/async_lib/async.lua:14: The coroutine failed with this message: ...im/site/pack/packer/start/gitsigns.nvim/lua/gitsigns.lua:402: attempt to index local 'bcache' (a nil value)
|| Error executing luv callback:
```

I don't have a reliable of reproducing it yet.

**Comment by lewis6991** [comment]
Thanks for reporting.

That error is happening in the index handler (the watcher that triggers when you update the git index). This can only happen if a buffer detaches (which clears the cache) but doesn't stop the watcher. A naive fix is simple by just aborting if there is no cache but that doesn't really address the root issue.

`nvim_buf_attach` seems to have some quirks that I don't fully understand yet so it's probably related to that.

**Comment by akinsho** [comment]
@lewis6991 I wonder if any of the following nvim PRs might relate:
https://github.com/neovim/neovim/pull/13895
https://github.com/neovim/neovim/pull/13935

I remember that at the time treesitter bugs were being fixed and some changes we're made to `buf_attach`'s behaviour or when it's called. Not sure if glancing at either might give you any clues (I personally have no idea).

**Comment by lewis6991** [comment]
Ah, there is now an `on_reload` callback. Will probably need to implement that in gitsigns. Thanks for the info, I'll need to look at those PRs in more detail.

**Comment by lewis6991** [comment]
Ah as I suspected, if no `on_reload` is provided (as current) then reload events just cause detach events instead. Adding a dummy `on_reload` callback prevents this.

**Comment by lewis6991** [comment]
Issue is probably the same but I've raised #139 .

**Comment by levouh** [comment]
I can reproduce this with reliably by doing:

1. Opening a buffer, edit it, write it, undo, edit again, write
2. Open the `fugitive` staging area with `:G`
3. Navigate back to the edited buffer and `:bd` it
4. `X` on the file in the staging area (this will `reset --hard HEAD` it)
5. Observe:

```
Error executing luv callback:
...acker/start/plenary.nvim/lua/plenary/async_lib/async.lua:14: The coroutine failed with this message: ...im/site/pack/packer/start/gitsigns.nvim/lua/gitsigns.lua:410: attempt to index local 'bcache' (a nil value)
```

although it looks like you've been working on this @lewis6991. Just figured it might be useful to throw this here as this still has the _unable to reproduce_ tag.

**Comment by joelpalmer** [comment]
With 462769d7a961042a7f2a49e2f57eb0e91467d260 everything was fixed for me as far as #139  and this. I'm on Neovim nightly from yesterday and plenary latest as well. Thanks for fixing this so quickly. Love this plugin. I saw it mentioned on Reddit in a reply to a question about gitgutter. Great work!
